timemachine/fe/single_topology.py

```diff
@@ -568,12 +654,73 @@ def assert_improper_idxs_are_canonical(all_idxs):
     )
 
 
+def find_chirally_valid_dummy_groups(
+    mol_a, mol_b, core: NDArray
+) -> Optional[Tuple[Dict[int, FrozenSet[int]], Dict[int, FrozenSet[int]]]]:
+    """Returns a pair of dummy group assignments for the A -> B and B -> A transformations such that the implied hybrid
+    mol is chirally valid, or None if no such pair exists.
+
+    Refer to :py:func:`check_chiral_validity` for definition of "chiral validity".
+
+    Refer to :py:func:`timemachine.fe.dummy.generate_dummy_group_assignments` and notes below for more
+    information on dummy group assignment.
+
+    For the A -> B transformation, dummy group indices refer to atoms in mol_b.
+    For the B -> A transformation, dummy group indices refer to atoms in mol_a.
+    """
+
+    bond_graph_a = convert_to_nx(mol_a)
+    bond_graph_b = convert_to_nx(mol_b)
+
+    # Use placeholder forcefield with a minimal number of patterns for performance
+    ff = Forcefield.load_from_file("placeholder_ff.py")
+
+    return find_chirally_valid_dummy_groups_impl(mol_a, mol_b, bond_graph_a, bond_graph_b, core, ff)
+
+
+def find_chirally_valid_dummy_groups_impl(
+    mol_a, mol_b, bond_graph_a: nx.Graph, bond_graph_b: nx.Graph, core: NDArray, ff: Forcefield
+) -> Optional[Tuple[Dict[int, FrozenSet[int]], Dict[int, FrozenSet[int]]]]:
+    """Implementation of :py:func:`find_chirally_valid_dummy_groups`, separated to allow precomputation of bond
+    graphs and forcefield."""
+
+    # TODO: Consider refactoring to also return harmonic bond and chiral atom potentials for the selected dummy groups,
+    # to avoid reconstructing later
+
+    with warnings.catch_warnings():
+        # Suppress warnings from end-state setup during the search; these are only relevant for the selected candidate,
+        # and will be raised again when we construct the final SingleTopology instance
+        warnings.simplefilter("ignore", ChiralVolumeDisabledWarning)
+        warnings.simplefilter("ignore", MultipleAnchorWarning)
+
+        pairs = (
+            (dummy_groups_ab, dummy_groups_ba)
+            # NOTE: We use fair_product_2 instead of itertools.product to avoid iterating over all possible candidates
+            # for A -> B before trying the second candidate for B -> A
+            for dummy_groups_ab, dummy_groups_ba in fair_product_2(
```

Comment:
[nit/tangent/curiosity] is the expected performance with `fair_product_2` ordering >= randomized ordering > `product` ordering?

(I don't quickly see the motivation for `fair_product_2` over a random permutation, but I see the motivation over the default ordering from `product`)

Reply:
At this stage I don't think there would be any benefit to using the fair product over a random permutation of the product (since `generate_dummy_group_assignments` returns an arbitrary ordering), but in principle the inputs could be ordered by some measure of optimality.

Reply:
Makes sense -- written in a modular way, could swap out different functions in place of `fair_product_2` if beneficial in the future